deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/layers/samediff/BaseSameDiffLayer.java

```diff
@@ -0,0 +1,67 @@
+package org.deeplearning4j.nn.conf.layers.samediff;
+
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import org.deeplearning4j.nn.api.ParamInitializer;
+import org.deeplearning4j.nn.conf.InputPreProcessor;
+import org.deeplearning4j.nn.conf.NeuralNetConfiguration;
+import org.deeplearning4j.nn.conf.inputs.InputType;
+import org.deeplearning4j.nn.conf.layers.Layer;
+import org.deeplearning4j.nn.conf.memory.LayerMemoryReport;
+import org.deeplearning4j.nn.layers.samediff.SameDiffLayer;
+import org.deeplearning4j.nn.params.SameDiffParamInitializer;
+import org.deeplearning4j.nn.weights.WeightInit;
+import org.deeplearning4j.optimize.api.IterationListener;
+import org.nd4j.autodiff.samediff.SDVariable;
+import org.nd4j.autodiff.samediff.SameDiff;
+import org.nd4j.linalg.api.ndarray.INDArray;
+import org.nd4j.linalg.learning.config.IUpdater;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+
+@Data
+@EqualsAndHashCode(callSuper = true)
+public abstract class BaseSameDiffLayer extends AbstractSameDiffLayer {
+
+    protected WeightInit weightInit;
+
+    protected BaseSameDiffLayer(Builder builder){
+        super(builder);
+        this.weightInit = builder.weightInit;
+    }
+
+    protected BaseSameDiffLayer(){
+        //No op constructor for Jackson
+    }
+
+    public abstract List<String> defineLayer(SameDiff sameDiff, SDVariable layerInput, Map<String,SDVariable> paramTable);
```

Comment:
I felt this would be better defined as a List<SDVariable> instead of List<String>.  That way you aren't just throwing away objects, and it's impossible to make a typo.

Reply:
Here's an idea oh how it could work (and a working unit test)

https://gist.github.com/wmeddie/809c602311503a2ebb2d2cd957daf09e